### ISSUE TITLE
[java] UnnecessaryParentheses fails to detect obvious scenario

### ISSUE THREAD

**Issue by jsotuyod** [pmd/pmd#537]
**Rule Set:**
java-controversial/UnnecessaryParentheses

**Description:**
The rule is currently too naive, only looking for parenthesis around a single term, but fails for several other trivial scenarios.

I'm not sure this will be achievable in XPath, we may have to convert the rule to a Java one....

**Code Sample demonstrating the issue:**

```java
public class Foo {
  private String getData(final String header) {
    String result = ("[" + header + ":" + Utils.decode(header) + "]");
    return result;
  }
}
```

**Running PMD through:** *Any*

**Comment by oowekyala** [comment]
I think I found a way to do that, transforming the XPath into:
```
         //Expression
           /PrimaryExpression
            /PrimaryPrefix
             /Expression[count(*)=1]
              /*[self::PrimaryExpression
                 or self::AdditiveExpression]
```
Adding `or self::CastExpression` and other types can allow us to match more incorrectly parenthesized expressions. Oddly though, there's a unit test for that rule which specifically says your example should be unreported:
```xml
    <test-code>
        <description>ok, complex expression</description>
        <expected-problems>0</expected-problems>
        <code><![CDATA[
        public class Foo {
          int bar() {
            return (2+2); 
          }
        }
        ]]></code>
    </test-code>
```

I think it's incorrect, can I delete it?

**Comment by jsotuyod** [comment]
By storing the value to a local var the AST is slightly different producing the false positive.

I wonder if simply doing:

```
    //Expression
           /PrimaryExpression
            /PrimaryPrefix
             /Expression[count(*)=1]
```

wouldn't work just fine...

**Comment by oowekyala** [comment]
Hum so the rule should not report parenthesized return statements ?

Your solution would fail on parenthesized cast expressions, like so: ` return ((Exception) o).getMessage();`

**Comment by jsotuyod** [comment]
damn! I'll have to tinker with it when I can... feel free to do so if you want to!

**Comment by oowekyala** [comment]
```
        //Expression[not(ancestor::ReturnStatement)]
           /PrimaryExpression[count(*)=1]
            /PrimaryPrefix
             /Expression[count(*)=1]
```
Your expression was a good start though! I changed it to that, which passes all existing tests

**Comment by oowekyala** [comment]
Hum actually it doesn't sorry x)

**Comment by oowekyala** [comment]
Well it appears that works:
```
         //Expression[ancestor::ReturnStatement]
           /PrimaryExpression[count(*)=1]
            /PrimaryPrefix
             /Expression[count(*)=1]
              /PrimaryExpression[count(*)=1]
          |
          //Expression[not(ancestor::ReturnStatement)]
           /PrimaryExpression[count(*)=1]
            /PrimaryPrefix
             /Expression[count(*)=1]
```

**Comment by oowekyala** [comment]
But if the rule really reported *unnecessary* parentheses I think it should report the `return (2+2)` example above

**Comment by jsotuyod** [comment]
@oowekyala I agree, I think we should treat `return (2+2)` as a false negative. Maybe we can look at the history (blame!) to understand better why that scenario was considered ok and see if the rationale is still valid.

**Comment by oowekyala** [comment]
Well the [blame](https://github.com/pmd/pmd/blame/master/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml) of the file is quite obscure... My 2 cents on this would be that parenthesizing the expression increases readibility, but ever so slightly. Compare `return (2 + 2);` and `return 2 + 2;`. I personally think it doesn't do much.

Also, I think the rule might fill better its contract if it reported unnecessary parentheses in situations like that:
```
return (2 + 2) + 3 + 4;
return ((String) "doo");
``` 
But as you said previously, I think it might be complicated to do that with XPath, as we probably would have to consider arithmetic expressions, cast expressions and primary expressions separately, as they have special cases

**Comment by jsotuyod** [comment]
@oowekyala yes, the blame is useless. This goes back way farther than expected.

We should just change it to whatever makes sense now then. All unnecessary parenthesis are to be reported. Please, submit a PR with [your suggested improvement](https://github.com/pmd/pmd/issues/537#issuecomment-321284375).

As we move on we may try to detect more scenarios.
